### ISSUE TITLE
Expose low-level telephony interface to support Gecko impl

### ISSUE THREAD

**Issue by joneschrisg** [andreasgal/B2G#27]
We need to strip out the android-Java stuff and get down to the C/C++ interface.  If it proves more expedient, we could create our own TelephonyProvider in Java and postpone the complete strip-out, but if possible it would be better do both in one shot.

I may not be able to own this.  I'll look through the code and try to analyze what's needed to do this, and post here.

**Comment by joneschrisg** [comment]
I made https://wiki.mozilla.org/TelephonyGlossary for acronyms I came across in the android source.  Hope it helps someone else.

**Comment by jaoo** [comment]
@mkedwards - Could you post a brief description about the status of this issue?

**Comment by ghost** [comment]
This is the repo I'm maintaining as a demo of how to strip this out: https://github.com/kmachulis-mozilla/b2g-dialer-test

**Comment by joneschrisg** [comment]
@kmachulis-mozilla: @jaoo is ready to help with the low-level implementation here.  Let's try to find some parallelizable work.  Here are the work items I know of
1. continue prototyping RIL interaction as you're doing here in the repo above.  Get the prototype to connect to rild's socket.
2. learn and prototype the RIL messages needed to initiate a phone call.  This can happen in parallel to (1), but (1) is required to test it.
3. prototype the protocol for sending an SMS message.  Testing also blocks on (1).
4. prototype accepting an incoming phone call.  Testing blocked on (1).
5. prototype receiving an SMS message.  Testing blocked on (1).
6. add C++ support for the android IPC mechanisms to Gecko.  That is, port the prototype from (1) to use gecko's event loop etc.
7. port the phone and SMS protocols to Gecko under our Telephony and SMS APIs.  That is, add them as a backend to Gecko.

For 2-5 above, I think Kyle's approach is good: write a minimal C program that can, e.g., dial a single phone number or receive an SMS message.  Kyle, will we need to disconnect the built-in dialer/SMS applications to make this work?

@jaoo, which items above would you like to work on?

**Comment by ghost** [comment]
Right now the big item is "figure out what we need to link to (library-wise) in the android IPC layer in order to build Parcels". Once we can build/send parcels, then we can start figuring out the flow of different events (calls, sms, etc...). I'll write up tasks on b2g-dialer-test and we can go from there.

**Comment by jaoo** [comment]
We have started prototyping RIL interaction and getting the prototype ready to connect to the rild's socket. Right now we are ready to play with the rild daemon and to send request to it. We are able to establish new connections but we have not sent any request yet. This program also links to some libraries such as libutils, libcutils, libbinder (needed to build Parcels) and libril.  

Tomorrow We are forking from b2g-dialer-test in order to push the code. From that point We can start working on any work item you posted above.

@kmachulis-mozilla, will you write up on b2g-dialer-test the set of task you mentioned?

**Comment by ghost** [comment]
@jaoo I just pushed my code, which is in pretty much the same place as yours. I got requests running last night, but it required actually changing the java (RIL.java) to try to connect to a different port (rild2 instead of rild), since the socket requires exclusive access. My next step was going to be writing an non-blocking I/O loop to service requests, because rild gets panicy if you drop /anything/, but it sounds like we should probably sync up.

I'm going to clean up my notes a bit and check those in too.

**Comment by ghost** [comment]
Ok, finally started putting in issues at

https://github.com/kmachulis-mozilla/b2g-dialer-test/issues

Quick note on my labelling scheme:
- Research means adding to documentation
- Task means adding to code
- Bug means fixing anything that's already there.

**Comment by ferjm** [comment]
@kmachulis-mozilla I've just forked b2g-dialer-test and pushed our connection test (RILfake_java.cpp) followed by some minor additions to your code. We could not make it work as we were receiving the "RILD must accept socket from radio" error message with any connection attempt, so we have included the switchUser() function in order to switch the test executable uid to radio. We've also needed to change the ril socket permissions to make it work.
